diff -Nur linux-2.6.22.old//net/ipv6/ip6_output.c linux-2.6.22/net/ipv6/ip6_output.c
--- linux-2.6.22.old//net/ipv6/ip6_output.c	2012-06-11 12:16:54.508897501 -0400
+++ linux-2.6.22/net/ipv6/ip6_output.c	2012-06-11 12:20:36.728897555 -0400
@@ -466,7 +466,8 @@
 		int addrtype = ipv6_addr_type(&hdr->saddr);
 
 		/* This check is security critical. */
-		if (addrtype & (IPV6_ADDR_MULTICAST|IPV6_ADDR_LOOPBACK))
+		if (addrtype == IPV6_ADDR_ANY ||
+		    addrtype & (IPV6_ADDR_MULTICAST|IPV6_ADDR_LOOPBACK))
 			goto error;
 		if (addrtype & IPV6_ADDR_LINKLOCAL) {
 			icmpv6_send(skb, ICMPV6_DEST_UNREACH,
