diff -Nur linux-2.6.22.old//include/linux/sysctl.h linux-2.6.22/include/linux/sysctl.h
--- linux-2.6.22.old//include/linux/sysctl.h	2012-06-11 12:26:26.878897639 -0400
+++ linux-2.6.22/include/linux/sysctl.h	2012-06-11 12:26:59.538897645 -0400
@@ -545,6 +545,7 @@
 	NET_IPV6_IP6FRAG_SECRET_INTERVAL=24,
 	NET_IPV6_MLD_MAX_MSF=25,
 	NET_IPV6_6TO4_FORCE_IP4FRAGOFF_ZERO=26,
+	NET_IPV6_PING6_INTERFACE=27,
 };
 
 enum {
diff -Nur linux-2.6.22.old//include/net/ipv6.h linux-2.6.22/include/net/ipv6.h
--- linux-2.6.22.old//include/net/ipv6.h	2012-06-11 12:26:26.918897637 -0400
+++ linux-2.6.22/include/net/ipv6.h	2012-06-11 12:27:46.088897657 -0400
@@ -570,6 +570,7 @@
 extern int sysctl_ip6frag_time;
 extern int sysctl_ip6frag_secret_interval;
 extern int sysctl_ipv6_6to4_force_ip4fragoff_zero;
+extern int sysctl_ipv6_ping6_interface;
 
 extern const struct proto_ops inet6_stream_ops;
 extern const struct proto_ops inet6_dgram_ops;
diff -Nur linux-2.6.22.old//net/ipv6/raw.c linux-2.6.22/net/ipv6/raw.c
--- linux-2.6.22.old//net/ipv6/raw.c	2012-06-11 12:26:26.838897638 -0400
+++ linux-2.6.22/net/ipv6/raw.c	2012-06-11 12:29:13.678897677 -0400
@@ -59,6 +59,8 @@
 #include <linux/proc_fs.h>
 #include <linux/seq_file.h>
 
+int sysctl_ipv6_ping6_interface = 0;
+
 struct hlist_head raw_v6_htable[RAWV6_HTABLE_SIZE];
 DEFINE_RWLOCK(raw_v6_lock);
 
@@ -732,10 +734,15 @@
 		    ipv6_addr_equal(daddr, &np->daddr))
 			daddr = &np->daddr;
 
-		if (addr_len >= sizeof(struct sockaddr_in6) &&
-		    sin6->sin6_scope_id &&
-		    ipv6_addr_type(daddr)&IPV6_ADDR_LINKLOCAL)
-			fl.oif = sin6->sin6_scope_id;
+		if (sysctl_ipv6_ping6_interface) {
+			if (addr_len >= sizeof(struct sockaddr_in6) && sin6->sin6_scope_id)
+				fl.oif = sin6->sin6_scope_id;
+		} else {
+			if (addr_len >= sizeof(struct sockaddr_in6) &&
+					sin6->sin6_scope_id &&
+					ipv6_addr_type(daddr)&IPV6_ADDR_LINKLOCAL)
+				fl.oif = sin6->sin6_scope_id;
+		}
 	} else {
 		if (sk->sk_state != TCP_ESTABLISHED)
 			return -EDESTADDRREQ;
diff -Nur linux-2.6.22.old//net/ipv6/sysctl_net_ipv6.c linux-2.6.22/net/ipv6/sysctl_net_ipv6.c
--- linux-2.6.22.old//net/ipv6/sysctl_net_ipv6.c	2012-06-11 12:26:26.838897638 -0400
+++ linux-2.6.22/net/ipv6/sysctl_net_ipv6.c	2012-06-11 12:28:20.918897665 -0400
@@ -88,6 +88,14 @@
 		.mode           = 0644,
 		.proc_handler   = &proc_dointvec
 	},
+	{
+		.ctl_name       = NET_IPV6_PING6_INTERFACE,
+		.procname       = "ipv6_ping6_interface",
+		.data           = &sysctl_ipv6_ping6_interface,
+		.maxlen         = sizeof(int),
+		.mode           = 0644,
+		.proc_handler   = &proc_dointvec
+	},
 	{ .ctl_name = 0 }
 };
 
