diff -Nurp linux-2.6.31.orig/arch/mips/atheros/prom.c linux-2.6.31/arch/mips/atheros/prom.c
--- linux-2.6.31.orig/arch/mips/atheros/prom.c  2011-03-04 16:33:44.000000000 +0800
+++ linux-2.6.31/arch/mips/atheros/prom.c       2011-03-04 16:34:02.000000000 +0800
@@ -42,7 +42,7 @@ static int inline ar7240_use__image_cmdl
 
 void __init prom_init(void)
 {
-	int memsz = 0x2000000, argc = fw_arg0, i;
+	int memsz = 0x2000000, argc = 0, i;
 	char **arg = (char **)fw_arg1;
 
 	printk("flash_size passed from bootloader = %d\n", (int)fw_arg3);
