diff -Nur linux-2.6.30.old//drivers/usb/core/driver.c linux-2.6.30/drivers/usb/core/driver.c
--- linux-2.6.30.old//drivers/usb/core/driver.c	2012-03-27 04:58:00.717321604 -0400
+++ linux-2.6.30/drivers/usb/core/driver.c	2012-03-27 05:01:44.277321656 -0400
@@ -615,13 +615,14 @@
 		/*Don't attach 3G modems in modem devices list, even they may have storage interface
 		 * Apples.Wang of DNI added this at April 23, 2010 */
 		int modem_match=0;
+		struct usb_device *uudev;
+		int uudev_vid=0, uudev_pid=0;
+		uudev = interface_to_usbdev(intf);
+		uudev_vid = le16_to_cpu(uudev->descriptor.idVendor);
+		uudev_pid = le16_to_cpu(uudev->descriptor.idProduct);
 		//printk("drv->name:%s\n", drv->name);
-		if(!strcmp(drv->name, "usb-storage")){
-			int modem_ct=0, modem_vid=0, modem_pid=0, uudev_vid=0, uudev_pid=0;
-			struct usb_device *uudev;
-			uudev = interface_to_usbdev(intf);
-			uudev_vid = le16_to_cpu(uudev->descriptor.idVendor);
-			uudev_pid = le16_to_cpu(uudev->descriptor.idProduct);
+		if(!strcmp(drv->name, "usb-storage") || !strcmp(drv->name, "KC NetUSB General Driver")){
+			int modem_ct=0, modem_vid=0, modem_pid=0;
 			//printk("uudev, VenderId: 0x%x, ProductId: 0x%x\n",uudev_vid,uudev_pid);
 			while(1){
 				modem_vid = usb_modem_ids[modem_ct].idVendor;
@@ -629,13 +630,18 @@
 				if(!modem_vid && !modem_pid)
 					break;
 				if(modem_vid == uudev_vid && modem_pid == uudev_pid){
-					//printk("USB Wireless Modem device detected, VenderId: 0x%x, ProductId: 0x%x\n",modem_vid,modem_pid);
+					//printk("USB Wireless Modem device detected, VenderId: 0x%x, ProductId: 0x%x, DriverName:%s\n",modem_vid,modem_pid, drv->name);
 					modem_match = 1;
 					break;
 				}
 				modem_ct ++;
 			}
 		}
+		if(modem_match != 1 && !strcmp(drv->name, "KC NetUSB General Driver")){
+        	        if(uudev_vid == HUAWEI_VENDOR_ID || uudev_vid == ZTE_VENDOR_ID)
+                	        modem_match = 1;
+        	}
+
 		if(modem_match)
 			return 0;
 #endif
diff -Nur linux-2.6.30.old//drivers/usb/usb-modem-list.h linux-2.6.30/drivers/usb/usb-modem-list.h
--- linux-2.6.30.old//drivers/usb/usb-modem-list.h	2012-03-27 04:58:00.717321604 -0400
+++ linux-2.6.30/drivers/usb/usb-modem-list.h	2012-03-27 05:06:09.987321720 -0400
@@ -491,10 +491,11 @@
 	{ USB_DEVICE(DELL_VENDOR_ID, DELL_PRODUCT_5730_MINICARD_SPRINT) },	/* Dell Wireless 5730 Mobile Broadband EVDO/HSPA Mini-Card */
 	{ USB_DEVICE(DELL_VENDOR_ID, DELL_PRODUCT_5730_MINICARD_TELUS) },	/* Dell Wireless 5730 Mobile Broadband EVDO/HSPA Mini-Card */
 	{ USB_DEVICE(DELL_VENDOR_ID, DELL_PRODUCT_5730_MINICARD_VZW) }, 	/* Dell Wireless 5730 Mobile Broadband EVDO/HSPA Mini-Card */
-	{ USB_DEVICE(ANYDATA_VENDOR_ID, ANYDATA_PRODUCT_ADU_E100A) },	/* ADU-E100, ADU-310 */
-	{ USB_DEVICE(ANYDATA_VENDOR_ID, ANYDATA_PRODUCT_ADU_500A) },
+	{ USB_DEVICE(ANYDATA_VENDOR_ID, ANYDATA_PRODUCT_ADU_E100A) },   /* ADU-E100, ADU-310A, ADU-300A */
+        { USB_DEVICE(0x05c6, 0x1000) }, //AnyDATA ADU-500A & ADU-520A status before modeswitch
+        { USB_DEVICE(ANYDATA_VENDOR_ID, ANYDATA_PRODUCT_ADU_500A) }, //AnyDATA ADU-500A & ADU-520A after modeswitch
 	{ USB_DEVICE(ANYDATA_VENDOR_ID, ANYDATA_PRODUCT_ADU_620UW) },
-	{ USB_DEVICE(AXESSTEL_VENDOR_ID, AXESSTEL_PRODUCT_MV110H) },
+        { USB_DEVICE(AXESSTEL_VENDOR_ID, AXESSTEL_PRODUCT_MV110H) },    //MV110 - Supported in Russia
 	{ USB_DEVICE(ONDA_VENDOR_ID, ONDA_PRODUCT_MSA501HS) },
 	{ USB_DEVICE(ONDA_VENDOR_ID, ONDA_PRODUCT_ET502HS) },
 	{ USB_DEVICE(ONDA_VENDOR_ID, 0x0003) },
@@ -619,13 +620,14 @@
 	{ USB_DEVICE(QUALCOMM_VENDOR_ID, 0xf000)},
 	{ USB_DEVICE(QUALCOMM_VENDOR_ID, 0x9000)},// f000 -> 9000
 	{ USB_DEVICE(QUALCOMM_VENDOR_ID, 0x2001)},
-	{ USB_DEVICE(0x1e0e, 0xce16)},// 2001 -> ce16  05c6 -> 1e0e
+        { USB_DEVICE(0x1e0e, 0xce16)},// 2001 -> ce16  05c6 -> 1e0e, DWM-162-U5 status after modeswitch
 	{ USB_DEVICE(0x1e0e, 0xcefe)},// 2001 -> cefe  05c6 -> 1e0e
 	{ USB_DEVICE(QUALCOMM_VENDOR_ID, 0x1000)},
 	{ USB_DEVICE(0x1e0e, 0xf000)},
 	{ USB_DEVICE(0x1e0e, 0x9000)}, // f000 -> 9000
 	{ USB_DEVICE(0x1e0e, 0x9200)}, // f000 -> 9200
-	{ USB_DEVICE(0x2077, 0x1000)},
+        { USB_DEVICE(0x2077, 0x1000)}, //Dlink DWM-162 & DWM-162-U5 status before modeswitch
+        { USB_DEVICE(0x1e0e, 0xce17)}, //Dlink DWM-162 status after modeswitch
 	{ USB_DEVICE(0x1e0e, 0xcd17)}, // 2077 -> 1e0e  1000 -> cd17
 //	{ USB_DEVICE(OPTION_VENDOR_ID, OPTION_PRODUCT_SCORPION_MODEM) }, // 1000 -> 6901 05c6 -> 0af0
 //	{ USB_DEVICE(OPTION_VENDOR_ID, OPTION_PRODUCT_GT_MAX_READY) }, // 1000 > 6701  05c6 -> 0af0
@@ -775,6 +777,7 @@
 	{ USB_DEVICE(0x16d8, 0x700a) },
 	{ USB_DEVICE(0x16d8, 0xf000) },
 	{ USB_DEVICE(0x16d8, 0x6006) },
+	{ USB_DEVICE(0x16d8, 0x5543) }, // C-motech CNM-550
 	{ USB_DEVICE(0x198f, 0xbccd) },
 	{ USB_DEVICE(0x198f, 0x0220) },
 	{ USB_DEVICE(ZTE_VENDOR_ID, 0x0003) },
@@ -840,5 +843,14 @@
 	{ USB_DEVICE(0x1fac, 0x0131) }, // 0130 -> 0131
 	{ USB_DEVICE(0x20a6, 0x1105) },
 
+        { USB_DEVICE(0x22de, 0x6801) }, //WeTelecom WM-D200
+        { USB_DEVICE(0x1edf, 0x6003) }, //Withtel WMU-100A & Airplus MCD-800
+        { USB_DEVICE(0x1edf, 0x6004) }, //Airplus MCD-650
+
+        //Below is wimax devices
+        { USB_DEVICE(0x0525, 0xa4a5) }, //Yota One
+        { USB_DEVICE(0x1076, 0x7708) }, //Yota One
+        { USB_DEVICE(0x0525, 0xa4a2) }, //Yota One
+
 	{ } /* Terminating entry */
 };
