diff -Nur linux-2.6.30.old/drivers/usb/class//cdc-acm.c linux-2.6.30/drivers/usb/class//cdc-acm.c
--- linux-2.6.30.old/drivers/usb/class//cdc-acm.c	2012-03-27 05:07:32.507321740 -0400
+++ linux-2.6.30/drivers/usb/class//cdc-acm.c	2012-03-27 05:09:51.627321774 -0400
@@ -1181,7 +1181,8 @@
 	usb_set_intfdata(data_interface, acm);
 
 	usb_get_intf(control_interface);
-	tty_register_device(acm_tty_driver, minor, &control_interface->dev);
+	//tty_register_device(acm_tty_driver, minor, &control_interface->dev);
+	tty_register_device(acm_tty_driver, minor, NULL);
 
 	acm_table[minor] = acm;
 
