diff -Nur linux-2.6.30.old//drivers/net/rtl819x/rtl_nic.c linux-2.6.30/drivers/net/rtl819x/rtl_nic.c
--- linux-2.6.30.old//drivers/net/rtl819x/rtl_nic.c	2012-11-29 03:39:08.188065323 -0500
+++ linux-2.6.30/drivers/net/rtl819x/rtl_nic.c	2012-11-29 03:46:55.748073141 -0500
@@ -11,6 +11,7 @@
  *   re865x_nic.c: NIC driver for the RealTek 865*
  *
  */
+#define DNI_LINK_CHANGE_NOTIFY 1
 
 #define DRV_RELDATE		"Mar 25, 2004"
 #include <linux/config.h>
@@ -41,6 +42,10 @@
 #include <net/rtl/rtk_vlan.h>
 #endif
 
+#ifdef DNI_LINK_CHANGE_NOTIFY
+#include <linux/sched.h>
+#endif
+
 #if	defined(CONFIG_RTL8196_RTL8366) && defined(CONFIG_RTL_IGMP_SNOOPING)
 #undef	CONFIG_RTL_IGMP_SNOOPING
 #endif
@@ -253,6 +258,11 @@
 static void one_sec_timer(unsigned long task_priv);
 #endif
 
+//Add for checking port status
+#ifdef DNI_LINK_CHANGE_NOTIFY
+static void check_phy_change(void);
+#endif
+
 #ifdef CONFIG_RTL8196C_GREEN_ETHERNET
 static void power_save_timer(unsigned long task_priv);
 #endif
@@ -3854,6 +3864,34 @@
 int port_pwr_save_low = 0;
 #endif
 
+//Add for checking port status
+#ifdef DNI_LINK_CHANGE_NOTIFY
+static void check_phy_change()
+{
+	extern int rtk_port_phyStatus_get(uint32 port, uint32 *pLinkStatus, uint32 *pSpeed, uint32 *pDuplex);
+	uint32 linkStatus, speed, duplex;
+	static int first_time=1;
+	static unsigned char last_port_status=0;
+	unsigned char curr_port_status;
+	int i;
+	
+	curr_port_status = 0;
+	for(i=0; i<5 ; i++)
+	{
+		if( rtk_port_phyStatus_get(i,&linkStatus,&speed,&duplex) != SUCCESS)
+			printk("Failed to get port status!!\n");
+		else if(linkStatus == 1)
+			curr_port_status |= (0x01 << i);
+	}
+	if(first_time)
+		first_time = 0;
+	else if(curr_port_status != last_port_status)
+		printk("##Pay your attention, cable plug in or off [%02x]!!!\n",curr_port_status);
+
+	last_port_status = curr_port_status;
+}
+#endif  //#ifdef DNI_LINK_CHANGE_NOTIFY
+
 #ifdef CONFIG_RTL_8196E
 void	refine_phy_setting(void)
 {
@@ -4189,6 +4227,11 @@
 
 	mod_timer(&cp->expire_timer, jiffies + HZ);
 
+	//Add for checking port status
+#ifdef DNI_LINK_CHANGE_NOTIFY
+	check_phy_change();
+#endif
+
 	//spin_unlock_irqrestore(&cp->lock, flags);
 	local_irq_restore(flags);
 }
