diff -Nur linux-2.6.30.old//drivers/net/rtl819x/rtl_nic.c linux-2.6.30/drivers/net/rtl819x/rtl_nic.c
--- linux-2.6.30.old//drivers/net/rtl819x/rtl_nic.c	2012-11-29 06:25:42.028064129 -0500
+++ linux-2.6.30/drivers/net/rtl819x/rtl_nic.c	2012-11-29 06:26:49.228064187 -0500
@@ -2304,7 +2304,7 @@
 					/*igmp packet*/
 				#if defined (CONFIG_NETFILTER)
 					if(rtl_MulticastRxFilterOff(skb, 4) == true)
-						return 0;//filter by iptables
+					//	return 0;//filter by iptables
 				#endif
 					rtl_igmpMldProcess(nicIgmpModuleIndex, skb->data, pid, &vlanRelayPortMask);
 					vlanRelayPortMask=rtl865x_getVlanPortMask(vid) & vlanRelayPortMask & ((1<<RTL8651_MAC_NUMBER)-1);
diff -Nur linux-2.6.30.old//net/bridge/br_forward.c linux-2.6.30/net/bridge/br_forward.c
--- linux-2.6.30.old//net/bridge/br_forward.c	2012-11-29 06:25:42.268064129 -0500
+++ linux-2.6.30/net/bridge/br_forward.c	2012-11-29 06:27:19.598064117 -0500
@@ -585,7 +585,7 @@
 
 			//printk("%s:%d,br0SwFwdPortMask is 0x%x,multicastFwdInfo.fwdPortMask is 0x%x\n",__FUNCTION__,__LINE__,br0SwFwdPortMask,multicastFwdInfo.fwdPortMask);
 			#if defined (CONFIG_RTL_HARDWARE_MULTICAST)
-			if((skb->srcVlanId!=0) && (skb->srcPort!=0xFFFF))
+			if((skb->srcVlanId!=0) && (skb->srcPort!=0xFFFF) && (skb->srcVlanId!=RTL_WANVLANID))
 			{
 				/*multicast data comes from ethernet port*/
 				#ifdef CONFIG_RTK_VLAN_WAN_TAG_SUPPORT
diff -Nur linux-2.6.30.old//net/bridge/br_input.c linux-2.6.30/net/bridge/br_input.c
--- linux-2.6.30.old//net/bridge/br_input.c	2012-11-29 06:25:42.268064129 -0500
+++ linux-2.6.30/net/bridge/br_input.c	2012-11-29 06:27:58.748064114 -0500
@@ -780,7 +780,7 @@
 					{
 						skb->dev=origDev;
 						DEBUG_PRINT(" filter a pkt:%d %s:% \n", k, skb->dev->name, &(dev_net(skb->dev)->ipv4.iptable_filter->name[0]));
-						goto drop;
+					//	goto drop;
 					}
 					else
 					{
