diff -Nur linux-2.6.30.old//net/rtl/fastpath/fastpath_common.c linux-2.6.30/net/rtl/fastpath/fastpath_common.c
--- linux-2.6.30.old//net/rtl/fastpath/fastpath_common.c	2012-11-29 07:16:48.868063763 -0500
+++ linux-2.6.30/net/rtl/fastpath/fastpath_common.c	2012-11-29 07:16:59.948063760 -0500
@@ -31,7 +31,7 @@
 #if defined(CONFIG_RTL_NF_CONNTRACK_GARBAGE_NEW)
 static int	rtl_fp_gc_rx_threshold;
 #endif
-static int fast_nat_fw = 1;
+static int fast_nat_fw = 0;
 
 
 #if defined(IMPROVE_QOS) || defined(CONFIG_RTL_HW_QOS_SUPPORT)
