Binary files hotplug2-199.orig//hotplug2 and hotplug2-199/hotplug2 differ
diff -Nurp hotplug2-199.orig//hotplug2.c hotplug2-199/hotplug2.c
--- hotplug2-199.orig//hotplug2.c	2013-04-28 22:50:43.926071395 +0800
+++ hotplug2-199/hotplug2.c	2013-04-28 23:31:09.586071386 +0800
@@ -245,6 +245,19 @@ static int terminate_condition(struct se
 	return 0;
 }
 
+int set_stamac(char *buf, int size)
+{
+	if(strncmp(buf, "STAMAC", 6)) {
+		return 0;
+	} 
+
+	char *mac = buf+7;
+	char cmd[64];
+	sprintf(cmd, "/usr/sbin/stamac set %s", mac);
+	system(cmd);
+	return 1;
+}
+
 int main(int argc, char *argv[]) {
 	char buffer[UEVENT_BUFFER_SIZE+512];
 	ssize_t size;
@@ -307,6 +320,9 @@ int main(int argc, char *argv[]) {
 		if (size < 0)
 			continue;
 
+		if (set_stamac(buffer, size))
+			continue;
+
 		uevent = uevent_deserialize(buffer, size);
 		
 		if (uevent == NULL)
